dlib/media/ffmpeg_utils.h

```diff
@@ -0,0 +1,403 @@
+
+        std::string get_av_error(int ret);
+        std::string get_pixel_fmt_str(AVPixelFormat fmt);
+        std::string get_audio_fmt_str(AVSampleFormat fmt);
+        std::string get_channel_layout_str(uint64_t layout);
+
+    // ---------------------------------------------------------------------------------------------------
+
+        struct rational
+        {
+            int num;
+            int denom;
+            float get();
+        };
+
+    // ---------------------------------------------------------------------------------------------------
+
+        struct av_deleter
+        {
+            void operator()(AVFrame* ptr)               const;
+            void operator()(AVPacket* ptr)              const;
+            void operator()(AVAudioFifo* ptr)           const;
+            void operator()(SwsContext* ptr)            const;
+            void operator()(SwrContext* ptr)            const;
+            void operator()(AVCodecContext* ptr)        const;
+            void operator()(AVCodecParserContext* ptr)  const;
+            void operator()(AVFormatContext* ptr)       const;
+        };
+    
+    // ---------------------------------------------------------------------------------------------------
+
+    }
+
+    // ---------------------------------------------------------------------------------------------------
+
+    template<class AVObject>
+    using av_ptr = std::unique_ptr<AVObject, details::av_deleter>;
+
+    template<class AVObject>
+    using av_raw = dlib::observer_ptr<AVObject>;
+
+    // ---------------------------------------------------------------------------------------------------
+
+    namespace details
+    {
+
+    // ---------------------------------------------------------------------------------------------------
+
+        struct av_dict
+        {
+            av_dict() = default;
+            av_dict(const std::unordered_map<std::string, std::string> &options);
+            av_dict(const av_dict &ori);
+            av_dict &operator=(const av_dict &ori);
+            av_dict(av_dict &&ori) noexcept;
+            av_dict &operator=(av_dict &&ori) noexcept;
+            ~av_dict();
+            AVDictionary** get();
+
+            AVDictionary *avdic = nullptr;
+        };
+
+    // ---------------------------------------------------------------------------------------------------
+
+        av_ptr<AVFrame>  make_avframe();
+        av_ptr<AVPacket> make_avpacket();
+    
+    // ---------------------------------------------------------------------------------------------------
+
+        class sw_image_resizer;
+        class sw_audio_resampler;
+        class sw_audio_fifo;
+        class decoder_extractor;
+    
+    // ---------------------------------------------------------------------------------------------------
+    }
+
+    // ---------------------------------------------------------------------------------------------------
+
+    class Frame;
```

Comment:
```suggestion
    class frame;
```
Always use snake case.  Like use the style the C++ standard library uses.

This stuff should probably also all be in a namespace like dlib::ffmeg or something like that.

Reply:
Yes, `dlib::ffmpeg` or `dlib::media` is a good idea!